---
title: "Curate_phylogroup_data"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_merlin/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/')
```

# Environment

```{r environment, echo=T, message=FALSE, include=T, results=T}
#Packages
packages <- c("tidyverse") 

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T)) 

# Print environment
set.seed(45) 
Sys.info()
sessionInfo()
```
# Load and clean data

```{r}
# Load data
clermont_data <- read.delim("~/Desktop/gl_merlin/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/data/ecoli_phylogroup/ecoli_phylogroup_phylogroups.txt",header = F) 
colnames(clermont_data) <-  c("file","PCR_detected_genes","quadruplex_genotype",'quadruplex_ambiguity','ecoli_phylogroup',"mash_file")
clermont_data <- clermont_data %>% select(-mash_file)  

# Add genomic ID variable
clermont_data$genomic_id <- gsub("_contigs_l1000.fasta","",clermont_data$file)
clermont_data <- clermont_data %>% select(-file)
```

# Save Data
```{r}
saveRDS(clermont_data,"~/Desktop/gl_merlin/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/data/ecoli_phylogroup/ecoli_phylogroup_curated.RDS")
```
