---
title: "Generate AMRFinder + MOB-Suite linked data on gene presence"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = TRUE,warning = FALSE, message = FALSE) 
rm(list = ls())
opts_knit$set(root.dir='~/Desktop/gl_MERLIN/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/')
```

# Environment

```{r environment, echo=T, message=FALSE, include=T, results=T}
#Packages
packages <- c("tidyverse","kableExtra") 

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T)) 

# Print environment
set.seed(45) 
Sys.info()
sessionInfo()
```

# Load data

```{r}  
# Dataframe
df <-  readRDS("./data/dataset/MERLIN_cohort_metadata_final.RDS")  %>% subset(WGS_escre == TRUE & qc_pass == TRUE) 
isolates <- df$genomic_id

# AMRFinder
amrfinder <- readRDS("./data/amrfinder/amrfinder_cleaned.RDS")  %>% subset(genomic_id %in% isolates)
amrfinder_mat <- readRDS("./data/amrfinder/amrfinder_matrix.RDS") %>% subset(rownames(.) %in% isolates)
ESCrE_genotype_info <- readRDS("./data/amrfinder/beta_lactamase_summary_MERLIN_cohort.RDS")
 
# MOB-Suite
mobsuite_contigs <- readRDS("./data/mobsuite/mobsuite_contigs.RDS") %>% subset(genomic_id%in% isolates)
mobtyper <- readRDS("./data/mobsuite/mobtyper.RDS")%>% subset(genomic_id %in% isolates)
``` 

# Annotate contigs with AMRFinder data

```{r} 
genes <- ESCrE_genotype_info$element_symbol_renamed %>% unique %>% sort

annotate_contigs <- function(gene,amrfinder_df,mobsuite_contig_df){
  amrfinder_w_gene <- amrfinder_df %>% subset(element_symbol_renamed == gene)
  contigs_w_gene <- amrfinder_w_gene$contig_id %>% sort %>% unique
  pres <- ifelse(mobsuite_contig_df$contig_id %in% contigs_w_gene,1,0)
  row <- cbind.data.frame(pres) %>% `colnames<-`(c(gene)) 
  return(row)
}

gene_pres <- lapply(genes,FUN=annotate_contigs,amrfinder_df = amrfinder,mobsuite_contig_df = mobsuite_contigs) %>% do.call(cbind,.) %>% mutate(contig_id = mobsuite_contigs$contig_id)

mobsuite_contigs <- left_join(mobsuite_contigs,gene_pres)
```

# Binary variable indicating chromosmoe or plasmid as chromosome or plasmid
```{r}
mobsuite_contigs$chromosome <- ifelse(mobsuite_contigs$molecule_type == "chromosome", T, F)
mobsuite_contigs$plasmid <- ifelse(mobsuite_contigs$molecule_type == "plasmid", T, F) 
```

# Contig data
```{r}
count_chromosome <- function(isolate,genes,isolate_contigs){
  chromosome <- subset(isolate_contigs, molecule_type == "chromosome")
  chromosome_sum <- chromosome %>% select(genes) %>% summarise_all(sum)
  chromosome_sum_row <- cbind(genomic_id = isolate, type = "chromosome", chromosome_sum)
  return(chromosome_sum_row)
}

count_plasmid <- function(cluster,isolate,genes,isolate_contigs){
  plasmid <- subset(isolate_contigs, primary_cluster_id == cluster)
  plasmid_sum <- plasmid %>% select(genes) %>% summarise_all(sum)
  plasmid_sum_row <- cbind(genomic_id = isolate, type = cluster, plasmid_sum)
  return(plasmid_sum_row)
}

count_genes_mobsuite <- function(isolate,genes,mobsuite_contigs){
  isolate_contigs <- mobsuite_contigs %>% subset(genomic_id == isolate)
  # Chromsome
  chromosome_info <- count_chromosome(isolate,genes,isolate_contigs)
  # Plasmids
  plasmids <- isolate_contigs$primary_cluster_id %>% unique %>% subset(. !="-") %>% sort
  plasmid_info <- lapply(plasmids,FUN=count_plasmid,isolate=isolate,genes =genes,isolate_contigs=isolate_contigs) %>% do.call(rbind,.)
  # Overall info
  data <- rbind(chromosome_info,plasmid_info)
  return(data)
} 
qc_genomes <- amrfinder$genomic_id %>% unique %>% sort

mobsuite_gene_pres <- lapply(qc_genomes,FUN=count_genes_mobsuite,genes=genes,mobsuite_contigs=mobsuite_contigs) %>% do.call(rbind,.)
```

# Annotate w/ AMR + Blactam categories
```{r}
AMR_genes <- amrfinder %>% subset(type =="AMR") %>% .$element_symbol_renamed %>% unique %>% sort 
BLA_genes <- ESCrE_genotype_info$element_symbol_renamed %>% unique %>% sort
ESCrE_genes <- ESCrE_genotype_info %>% subset(ESCrE ==T) %>% .$element_symbol_renamed %>% unique %>% sort

mobsuite_gene_pres$AMR_contig <- ifelse(mobsuite_gene_pres %>% select(any_of(AMR_genes)) %>% rowSums()>0,T,F)
mobsuite_gene_pres$BETA_LACTAM_contig <- ifelse(mobsuite_gene_pres %>% select(any_of(BLA_genes)) %>% rowSums()>0,T,F) 
mobsuite_gene_pres$ESCrE_contig <- ifelse(mobsuite_gene_pres %>% select(any_of(ESCrE_genes)) %>% rowSums()>0,T,F) 

# Save the mobsuite gene presence data
saveRDS(mobsuite_gene_pres,"./data/ESBL_plasmids/amrfinder_mobsuite_beta_lactamase_presence.RDS")
```

# Get isolate's plasmid clusters with BLA and ESCrE genes
```{r} 
plasmid_sum_info_blactam <- mobsuite_gene_pres %>% subset(BETA_LACTAM_contig ==T)
plasmid_sum_info_ESCrE <- mobsuite_gene_pres %>% subset(ESCrE_contig ==T)

get_mobtyper_cluster_from_amrfinder <- function (df, mobtyper) 
{
    plasmid_cluster <- mobtyper$type %>% unique
    lapply(plasmid_cluster, FUN = function(x, df, mobtyper) {
        isolates_with_plasmid <- subset(mobtyper, type == 
            x) %>% select(genomic_id) %>% unlist
        row <- assign(paste0(x), ifelse(df$genomic_id %in% isolates_with_plasmid, 
            1, 0))
        return(row)
    }, df, mobtyper) %>% do.call(bind_cols, .) %>% as.data.frame %>% 
        `colnames<-`(plasmid_cluster) %>% `rownames<-`(df$genomic_id)
}

# B-lactamase gene contig matrix
## Implementation
blactam_plasmid_contigs <- get_mobtyper_cluster_from_amrfinder(df,plasmid_sum_info_blactam)
## Save
saveRDS(blactam_plasmid_contigs,"./data/ESBL_plasmids/amrfinder_beta_lactam_contig_matrix.RDS")

# ESCrE gene contig matrix
## Implementation
ESCrE_plasmid_contigs <- get_mobtyper_cluster_from_amrfinder(df,plasmid_sum_info_ESCrE)
## Save
saveRDS(ESCrE_plasmid_contigs,"./data/ESBL_plasmids/amrfinder_ESCrE_contig_matrix.RDS")
```
