---
title: "Supplemental Table 1: Location of ESBL genes on short-read assemblies"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_merlin/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/')
```

# Environment

```{r environment, echo=T, message=FALSE, include=T, results=T}
#Packages
packages <- c("tidyverse",'kableExtra') 

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

# Functions
source("./lib/consistent_themes.R")

# Print environment
set.seed(45) 
Sys.info()
sessionInfo()
```

# Data
```{r cars} 
ESCrE_genotype_data <- readRDS("./data/ESBL_plasmids/MERLIN_index_cohort_ESCrE_proportion_plasmidic.RDS")
ESCRE_genotype_info <- readRDS("./data/amrfinder/beta_lactamase_summary_MERLIN_cohort.RDS")
```

# Curate table
```{r}
ESCrE_genotype_data <- left_join(ESCrE_genotype_data %>% mutate(element_symbol_renamed = ESCrE),ESCRE_genotype_info %>% select(element_symbol_renamed,blactam_class,ESBL,carbapenemase,ampC_like))
ESCrE_genotype_data$category <- ifelse(ESCrE_genotype_data$ESBL == T,"ESBL",ifelse(ESCrE_genotype_data$carbapenemase == T,"Carbapenemase",ifelse(ESCrE_genotype_data$ampC_like ==T, "ampC-like","Other")))

ESBL_prop_data <- ESCrE_genotype_data %>% select(element_symbol_renamed,category,n_chromosome,n_plasmid,prop_plasmid) %>% arrange(-prop_plasmid,-n_plasmid,n_chromosome)

colnames(ESBL_prop_data) <- c("ESCrE","Category","No. chromosomal","No. plasmidic","Proportion plasmidic")
 
ESBL_prop_data %>% favorite_kable() 
```

# Save table
```{r}
saveRDS(ESBL_prop_data,"./tables/s_table_1_ESBL_proportion_plasmidic.RDS")
write_csv(ESBL_prop_data,"./tables/s_table_1_ESBL_proportion_plasmidic.csv")
```