---
title: "Summarise amrfinder results"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = TRUE,warning = FALSE, message = FALSE) 
rm(list = ls())
opts_knit$set(root.dir='~/Desktop/gl_MERLIN/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/')
```

# Environment

```{r environment, echo=T, message=FALSE, include=T, results=T}
#Packages
packages <- c("tidyverse","kableExtra") 

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T)) 

# Print environment
set.seed(45) 
Sys.info()
sessionInfo()
```

# Load data
```{r}
df <- readRDS("./data/dataset/MERLIN_cohort_metadata_final.RDS") 
isolates <- subset(df,WGS_escre == TRUE & qc_pass == TRUE) %>% .$genomic_id 
amrfinder <- readRDS("./data/amrfinder/amrfinder_cleaned.RDS")  %>% subset(genomic_id %in% isolates)
amrfinder_mat <- readRDS("./data/amrfinder/amrfinder_matrix.RDS") %>% subset(rownames(.) %in% isolates)
```

```{r}
blactam_info <-  amrfinder %>% subset(class == "BETA-LACTAM") %>% select(element_symbol,element_symbol_renamed,element_name,class,subclass) %>% distinct
paste0("Detected b-lactamase genes: ")
sort(blactam_info$element_symbol_renamed)

paste0("Detected ESBL genes: ")
ESBL_genes <- amrfinder[grepl("extended-spectrum",amrfinder$element_name),'element_symbol_renamed'] %>% unique %>% sort
sort(ESBL_genes)

paste0("Summary information about these beta-lactamases")

# Presence in isolates
blactam_info$count <- sapply(blactam_info$element_symbol_renamed, FUN=function(x){
  amrfinder_mat[[x]] %>% sum
})
blactam_info$proportion <- round(blactam_info$count / nrow(amrfinder_mat) * 100,2) 

# Info about genes of interest
blactam_info$blactam_class <- ifelse(grepl('class A',blactam_info$element_name),"A",
                                     ifelse(grepl('class B',blactam_info$element_name),"B",
                                            ifelse(grepl('class C',blactam_info$element_name),"C",
                                                   ifelse(grepl('class D',blactam_info$element_name),"D","NA"))))

# ESBL 
blactam_info$ESBL <- ifelse(blactam_info$element_symbol_renamed %in% ESBL_genes,TRUE,FALSE)

# Carbapenemase
BLA_carbapenem_genes <- amrfinder %>% subset(class =="BETA-LACTAM" & subclass=="CARBAPENEM") %>% .$element_symbol_renamed %>% unique %>% sort 
blactam_info$carbapenemase <- ifelse(blactam_info$element_symbol_renamed %in% BLA_carbapenem_genes,T,F)

# ampC_like
blactam_info$ampC_like <- ifelse(blactam_info$blactam_class =="C",T,F)

# Any ESCRE genotype
blactam_info$ESCrE <- ifelse(blactam_info$ESBL ==T | blactam_info$carbapenemase == T | blactam_info$ampC_like ==T, T,F)

paste0("ESCrE genotypes:")
blactam_info %>% subset(ESCrE ==T) %>% .$element_symbol_renamed %>% unique %>% sort

paste0("Overall data:")

blactam_info %>% select(-class) %>%  arrange(element_symbol) %>% gontjesr::favorite_kable()

# Save
blactam_info <- blactam_info %>% arrange(element_symbol_renamed)

```

# Save
```{r}
saveRDS(blactam_info,file="./data/amrfinder/beta_lactamase_summary_MERLIN_cohort.RDS")
```

