---
title: "Results Section 3: Sharing of ESBLs provides additional support for prolonged colonization"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_merlin/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/')
```


```{r environment, echo=F, message=FALSE, include=F, results=F}
#Packages
packages <- c("tidyverse",'ape','kableExtra','phytools','igraph','ggraph','tableone','kableExtra','ComplexHeatmap','gontjesr') 

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

# Source library
source("./lib/consistent_themes.R")

# Print environment
set.seed(45) 
Sys.info()
sessionInfo()
```

# Load data
```{r } 
df_index <- readRDS("./data/dataset/MERLIN_cohort_index_cases_metadata_final.RDS") 
prolonged_colonization_df <- readRDS("./data/prolonged_colonization/MERLIN_prolonged_colonization_index_patients.RDS")
plasmid_data <- readRDS("./data/ESBL_plasmids/MERLIN_beta_lactam_presence_location_data.RDS")

df_index <- left_join(df_index,plasmid_data)
```

```{r}
ESCrE_index_positive_households <- df_index %>% subset(visit ==0 & WGS_escre==T) %>% .$household
paste0("Number of subjects with index specimens positive: ", length(ESCrE_index_positive_households))
paste0("Proportion of subjects with index specimens positive: ", length(ESCrE_index_positive_households) / 40)
```


# Two case patients had episode of shraing ESBL plasmidic genes

```{r}
paste0("Strain vs. Plasmid sharing")
table(prolonged_colonization_df$index_at_1_and_or_2_Species_ST_cluster,prolonged_colonization_df$index_at_1_and_or_2_ESBL_plasmidic_genes,dnn=c("Strain","Plasmidic ESBL"))  

convert_tableone_into_df(dataset=prolonged_colonization_df,vars = "index_at_1_and_or_2_ESBL_plasmidic_genes",strata= "index_at_1_and_or_2_Species_ST_cluster",factorVars = "index_at_1_and_or_2_ESBL_plasmidic_genes",outcome_names = c("Not same strain","Same strain"))
```

```{r,results='asis'}
same_ESBL_not_same_strain <- prolonged_colonization_df %>% subset(index_at_1_and_or_2_Species_ST_cluster==0 & index_at_1_and_or_2_ESBL_plasmidic_genes==1) %>% .$household

lapply(same_ESBL_not_same_strain,FUN=function(x){
  
  df_index %>% subset(household==x) %>% select(household,genomic_id,Species_ST_cluster,visit,plasmids,bla_genes,ESBL_genes,ESBL_plasmidic_genes,  bla_chromosome_genes,ESBL_chromosome_genes,ESBL_plasmid_genes,bla_plasmids,ESBL_plasmids) %>% arrange(visit)   
}) %>% lapply(.,gontjesr::favorite_kable)
```

