---
title: "Results Section 4: Sharing of ESCrEs provides additional support for prolonged colonization"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_merlin/Analysis/Household_transmission/MERLIN-prolonged-colonization-ms/')
```

# Environment

```{r environment, echo=T, message=FALSE, include=T, results=T}
#Packages
packages <- c("tidyverse",'kableExtra','tableone','gontjesr')

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

# Source library
source("./lib/consistent_themes.R")

# Print environment
set.seed(45) 
Sys.info()
sessionInfo()
```

# Load data
```{r } 
df_index <- readRDS("./data/dataset/MERLIN_cohort_index_cases_metadata_final.RDS") 
prolonged_colonization_df <- readRDS("./data/prolonged_colonization/MERLIN_prolonged_colonization_index_patients.RDS")
plasmid_data <- readRDS("./data/ESBL_plasmids/MERLIN_index_cohort_beta_lactam_location_data.RDS")

df_index <- left_join(df_index,plasmid_data)
```

# Get list of index positive households
```{r}
ESCrE_index_positive_households <- df_index %>% subset(visit ==0 & WGS_escre==T) %>% .$household
paste0("Number of subjects with index specimens positive: ", length(ESCrE_index_positive_households))
paste0("Proportion of subjects with index specimens positive: ", length(ESCrE_index_positive_households) / 40)
```

# Three case patients had episode of shraing ESBL plasmidic genes

```{r}
paste0("Strain vs. ESCrE gene sharing")
table(prolonged_colonization_df$index_at_1_and_or_2_Species_ST_cluster,prolonged_colonization_df$index_at_1_and_or_2_ESCrE_genes,dnn=c("Strain","Plasmidic ESCrE"))  

convert_tableone_into_df(dataset=prolonged_colonization_df,vars = "index_at_1_and_or_2_ESCrE_genes",strata= "index_at_1_and_or_2_Species_ST_cluster",factorVars = "index_at_1_and_or_2_ESCrE_genes",outcome_names = c("Not same strain","Same strain"))
```

# How many were linked by plasmid?
```{r}
paste0("Strain vs. Plasmid-linked ESCrE genes")
table(prolonged_colonization_df$index_at_1_and_or_2_Species_ST_cluster,prolonged_colonization_df$index_at_1_and_or_2_ESCrE_genes_plasmid_aware,dnn=c("Strain","Plasmid-linked ESCrE"))  

paste0("ESCrE vs. Plasmid-linked ESCrE genes")
table(prolonged_colonization_df$index_at_1_and_or_2_ESCrE_genes,prolonged_colonization_df$index_at_1_and_or_2_ESCrE_genes_plasmid_aware,dnn=c("ESCrE genes","Plasmid-linked ESCrE"))  
```

## Nine linked to plasmid
```{r,results='asis'}
plasmid_linked_ESCrE_ <- prolonged_colonization_df %>% subset(index_at_1_and_or_2_ESCrE_genes==1 & index_at_1_and_or_2_ESCrE_genes_plasmid_aware==1) %>% .$household

lapply(plasmid_linked_ESCrE_,FUN=function(x){
  
  df_index %>% subset(household==x) %>% select(household,genomic_id,Species_ST_cluster,visit,plasmids,bla_genes,ESCrE_genes,ESCrE_plasmidic_genes,  bla_chromosome_genes,ESCrE_chromosome_genes,ESCrE_plasmid_genes,bla_plasmids,ESCrE_plasmids) %>% arrange(visit)   
}) %>% lapply(.,gontjesr::favorite_kable)
```

## Ten ESCrE sharing, but not linked to plasmid
```{r,results='asis'}
not_plasmid_linked_ESCrE_ <- prolonged_colonization_df %>% subset(index_at_1_and_or_2_ESCrE_genes==1 & index_at_1_and_or_2_ESCrE_genes_plasmid_aware==0) %>% .$household

lapply(not_plasmid_linked_ESCrE_,FUN=function(x){
  
  df_index %>% subset(household==x) %>% select(household,genomic_id,Species_ST_cluster,visit,plasmids,bla_genes,ESCrE_genes,ESCrE_plasmidic_genes,  bla_chromosome_genes,ESCrE_chromosome_genes,ESCrE_plasmid_genes,bla_plasmids,ESCrE_plasmids) %>% arrange(visit)   
}) %>% lapply(.,gontjesr::favorite_kable)
```

# Summary: Four cases where prolonged colonization at species-level found, but not strain-level

## One case was different sequence type but no shared ESCrE gene (i.e., likely acquisition)
```{r,results='asis'}
same_ESCrE_species_not_same_strain_no_ESBL_sharing <- prolonged_colonization_df %>% subset(index_at_1_and_or_2_Species_ST_cluster==0 & index_at_1_and_or_2_ESCrE_genes==0 &index_at_1_and_or_2_Species==1) %>% .$household

lapply(same_ESCrE_species_not_same_strain_no_ESBL_sharing,FUN=function(x){
  df_index %>% subset(household==x) %>% select(household,genomic_id,Species_ST_cluster,visit,plasmids,bla_genes,ESCrE_genes,ESCrE_plasmidic_genes,  bla_chromosome_genes,ESCrE_chromosome_genes,ESCrE_plasmid_genes,bla_plasmids,ESCrE_plasmids) %>% arrange(visit)   
}) %>% lapply(.,gontjesr::favorite_kable)
```

## Example cases explained by ESCrE but not strain

### Novel strain acquisition:
#### 1. 100H - Different E. coli ST - Linked by blaEC-8, but this was on the chromosome of both genomes...
#### 2. 169H- Different E. coli ST 131 clusters - Linked by CTX-M-15 and blaEC-5*. While blaEC-5 was chromosomal, CTX-M-15 was found on plasmid in 1 isolate, suggestive of HGT and/or co-colonization... 

### Plasmid sharing:
#### 1. 1H - Initial colonized with E. coli ST69 with blaCTX-M-15 and blaEC-8*. While blaEC-8 was found in subsequent chromosome of E. coli ST6321 strain, blaCTX-M-15 was found on a plasmid in every strain collected!!

```{r,results='asis'}
same_ESCrE_not_same_strain <- prolonged_colonization_df %>% subset(index_at_1_and_or_2_Species_ST_cluster==0 & index_at_1_and_or_2_ESCrE_genes==1) %>% .$household

lapply(same_ESCrE_not_same_strain,FUN=function(x){
  
  df_index %>% subset(household==x) %>% select(household,genomic_id,Species_ST_cluster,visit,plasmids,bla_genes,ESCrE_genes,ESCrE_plasmidic_genes,  bla_chromosome_genes,ESCrE_chromosome_genes,ESCrE_plasmid_genes,bla_plasmids,ESCrE_plasmids) %>% arrange(visit)   
}) %>% lapply(.,gontjesr::favorite_kable)
```
 